js/ui/codeView.js

```diff
@@ -1187,6 +1187,7 @@ var CodeViewManager = new Lang.Class({
     },
 
     _addSession: function(actor) {
+        actor.connect('destroy', this.handleDestroyWindow.bind(this));
```

Comment:
I have two main questions about this patch:
- why would we not be interested in the toolbox actor destroy events as well?
- more importantly, what happens when a window is unminimized? Consider the use case where a hackable app minimized, then unminimized again (or to add more complexity, even flipped before the first minimization, then flipped back). Would we not end up without a corresponding CodingSession for that hackable app/toolbox, with this patch?

Reply:
About the toolbox actor, I think it's not needed because the only actor that can be destroyed when minimized is the app actor.

This change don't remove the coding session on minimize, it only binds the destroy callback to the actor destroy event for the case when the shell doesn't notify the codeviewmanager about a window being destroyed (that's the case when it's minimized). So the only problem with this is the callback being called twice, but in that case, the callback won't do anything because the condingsession won't be there.

Reply:
> About the toolbox actor, I think it's not needed because the only actor that can be destroyed when minimized is the app actor.

I'm not sure what you mean by this, do you see that the application actor emits a destroy signal when minimized, but the toolbox actor does not do that?

> This change don't remove the coding session on minimize, it only binds the destroy callback to the actor destroy event for the case when the shell doesn't notify the codeviewmanager about a window being destroyed (that's the case when it's minimized). So the only problem with this is the callback being called twice, but in that case, the callback won't do anything because the condingsession won't be there.

Now I am a bit confused here. I thought that the issue is that upon minimization, the window actor is destroyed, but we were keeping the CodingSession alive. So your fix here connects to that destroy signal to remove the CodingSession when that happens. What is the purpose of connecting the callback if not to remove the session? (And you also say it yourself, the second time the callback is called it won't do anything because the session was already removed.)

Reply:
> > About the toolbox actor, I think it's not needed because the only actor that can be destroyed when minimized is the app actor.
> 
> I'm not sure what you mean by this, do you see that the application actor emits a destroy signal when minimized, but the toolbox actor does not do that?

What I'm saying is that when you close when minimized, the app actor will emit the destroy signal, so it's not needed to listen to the destroy signal of the toolbox to remove the coding session, the actor destroy signal will do the work.

> > This change don't remove the coding session on minimize, it only binds the destroy callback to the actor destroy event for the case when the shell doesn't notify the codeviewmanager about a window being destroyed (that's the case when it's minimized). So the only problem with this is the callback being called twice, but in that case, the callback won't do anything because the condingsession won't be there.
> 
> Now I am a bit confused here. I thought that the issue is that upon minimization, the window actor is destroyed, but we were keeping the CodingSession alive. So your fix here connects to that destroy signal to remove the CodingSession when that happens. What is the purpose of connecting the callback if not to remove the session? (And you also say it yourself, the second time the callback is called it won't do anything because the session was already removed.)

The real problem is when we close the app when it's minimized. Minimize and unminimize doesn't affect at all and works correctly.

But if you minimize and then right click and close the app, when it's minimized, the actual code doesn't call to the handleDestroyWindow, the shell wm doesn't emit the "destroy" signal, so we have a codingSession in the list pointing to a destroyed app actor and when we try to do "session.app" the shell crashes.

That's why listening to the actor destroy signal fixes this, because that signal is emitted always, no matter how the actor is being destroyed.

Reply:
Again apologies that this took so long to reply to!

> What I'm saying is that when you close when minimized, the app actor will emit the destroy signal, so it's not needed to listen to the destroy signal of the toolbox to remove the coding session, the actor destroy signal will do the work.

I would expect that to depend on whether the app or the toolbox is "in front". Due to `_applyWindowMinimizationState()` when one window is minimized, both are.  But e.g. in the overview window picker and on the taskbar menu, we always show the window in front, to reinforce the illusion that those are two sides of the same window.

So if you e.g. minimize the app while the toolbox side is showing, then close it by clicking on the X next to it in the window picker, I expect the destroy signal to reach the toolbox actor, not the app one.

> The real problem is when we close the app when it's minimized. Minimize and unminimize doesn't affect at all and works correctly.
> 
> But if you minimize and then right click and close the app, when it's minimized, the actual code doesn't call to the handleDestroyWindow, the shell wm doesn't emit the "destroy" signal, so we have a codingSession in the list pointing to a destroyed app actor and when we try to do "session.app" the shell crashes.
> 
> That's why listening to the actor destroy signal fixes this, because that signal is emitted always, no matter how the actor is being destroyed.

Thanks, things make a lot more sense after your explanation and I now understand why you prefer this patch over the other PR! :-)
I didn't realize that we would not receive a signal from the WindowManager class while windows are minimized.

Would it be possible for you to change this patch so that the Clutter destroy signal only stays connected during the time we know we won't get one from the WM? (e.g. only between the time the window is minimized and unminimized)
I think that would address the last concern I have about this.

Reply:
I've updated the code to only listen to the destroy event when the window is minimized. I needed to add a signal to be able to know if the `CodingSession` actor is minimized in the `CodeViewManager` because we need to remove the `CodingSession` from the `CodeViewManager` object.